### DISCUSSION TITLE
How to achieve this spline and particle effect?

### DISCUSSION THREAD

**Discussion by Dumanbokaei** [JacquesLucke/animation_nodes#1685] [Q&A] [answered]
How to achieve the following effect in Blender?
![image](https://user-images.githubusercontent.com/33803084/112281400-0f534d80-8c97-11eb-9656-6b1465d7beca.png)

[Source](https://www.youtube.com/watch?v=S7wOVM6HaWE)

I tried to apply curves to a deformed surface, but that was a bad idea. the distribution of the curves looks like a magic texture, but working with such didn't work either.

Is this something that can be done with Animation Nodes? Or with a particle system?

**Comment by Gerstmann-Bradley** [comment, marked as answer]
Definitely doable with animation nodes. (You may also need basic shader node)
I don't have much time today. Will have to answer this question like at least half-day later.

Basic Idea I will go is to create splines using loop. Within the loop each spline can be deformed differently, and transformed by matrix from circle (distribute matrices - circle mode).
You can trim spline for animation.
also evaluate spline to instance all these small particles..
**********************

https://user-images.githubusercontent.com/55994236/112628672-958fa100-8e09-11eb-9fe0-d8057dcf42d0.mp4
I have made a simple example with my own presets.
For convenience, I will only show the most basic set up as a start point, instead of the actual setup used to produce this animation.

The basic principle has been explained earlier already. As for details: 
1. Create Splines Within a Loop (Green loop).
     a) The upper matrix list input is to define location, which has been used to transform the spline as discussed later. 
     b) The lower vector list input is to set a basic model of spline shape, which later will be randomized within the loop. Here I simply use default Shift+A "beizer curve" as the basic model.
![image](https://user-images.githubusercontent.com/55994236/112629734-0e432d00-8e0b-11eb-9590-74d58a294de6.png)
2. Within the loop, randomize vector list, and form new spline. Set spline radius using float range so it's thick at the beginner and get thinner later.
3. Transform the spline using circle matrix. output the spline, and spline's evaluation which can be used for particle location.
![image](https://user-images.githubusercontent.com/55994236/112629036-26667c80-8e0a-11eb-96ae-d43fc9322187.png)
4. Outside loop, generate spline and instances.
![image](https://user-images.githubusercontent.com/55994236/112629960-582c1300-8e0b-11eb-8dc0-f39f973e0cdf.png)

At the end:
As we can't directly keyframe within animation nodes. It's recommend use "object transform input" to keyframe.
Additional functionality needs to be added on the top of this node tree according to your specific needs, which is too complicate to explain, I think I will stop here. And specific animation definitely includes various keyframes.

**Reply by Dumanbokaei** [reply]
Thanks for the response. There's no rush. I'll wait for your reply.

**Reply by Gerstmann-Bradley** [reply]
(Answers has been moved to the initial comment.)

**Reply by OmarEmaraDev** [reply]
Thanks for answering @Gerstmann-Bradley! You should probably add this as an answer instead of a comment.

**Reply by Gerstmann-Bradley** [reply]
> Thanks for answering @Gerstmann-Bradley! You should probably add this as an answer instead of a comment.

I see, I suppose you mean the most initial comment. If so I have just done that. Thanks for mentioning.

**Reply by OmarEmaraDev** [reply]
@Gerstmann-Bradley  Ah, I didn't realize your comment was itself an answer. In this case, in the future, we should probably just add a new answer and mark it as an answer like I just did. Is everybody allowed to mark answers or is it just for maintainers?

**Reply by Gerstmann-Bradley** [reply]
> @Gerstmann-Bradley Ah, I didn't realize your comment was itself an answer. In this case, in the future, we should probably just add a new answer and mark it as an answer like I just did. Is everybody allowed to mark answers or is it just for maintainers?

I'm not familiar with Github about this. I don't think I can mark my comments as answer by my own. Perhaps questioner can select the answer as well as the administrator, but I'm not sure.
![image](https://user-images.githubusercontent.com/55994236/112634842-785ed080-8e11-11eb-819f-e411825773a6.png)

**Reply by Dumanbokaei** [reply]
@OmarEmaraDev  & Bradly, thank you both for your answers. It was great. I'll try it and come back to you if there's any problem.

**Comment by OmarEmaraDev** [comment]
Here is a tip from you which you can implement on top of @Gerstmann-Bradley's setup for more control. Consider this node tree for simplicity:

![20210326-225625](https://user-images.githubusercontent.com/25300994/112691580-33c14c80-8e86-11eb-86f3-cc4e4ad35ceb.png)

If you evaluate the angles at a noise then add the output to the angle, the angles will have varying degrees of compression, achieving the clumping you see in your example:

![20210326-230112](https://user-images.githubusercontent.com/25300994/112691867-9e728800-8e86-11eb-804c-2fc7637466e9.png)

This can be applied on the individual control points of the spline for more control. You can also modulate the amplitude and frequency of the noise based on the distance from the center for even better control, but I will leave that for you.
